graders/c/cgrader/cgrader.py

```diff
@@ -302,69 +301,90 @@ def test_run(
         elif name is None:
             name = "Test command: %s" % command
 
-        if exp_output is not None and not isinstance(exp_output, list):
+        if exp_output is None:
+            exp_output = []
+            must_match_all_outputs = True
+        elif not isinstance(exp_output, list):
             exp_output = [exp_output]
-        if reject_output is not None and not isinstance(reject_output, list):
+
+        if reject_output is None:
+            reject_output = []
+        elif not isinstance(reject_output, list):
             reject_output = [reject_output]
-        if msg is None and exp_output is not None:
-            msg = "Expected: %s" % (" AND " if must_match_all_outputs else " OR ").join(
-                [f"\n{t}\n" if "\n" in str(t) else f'"{t}"' for t in exp_output]
-            ) + (
-                ' but not "%s"' % '"/"'.join([str(t) for t in reject_output])
-                if reject_output
-                else ""
+
+        def compile_re(t):
+            if isinstance(t, re.Pattern):
+                return (t.pattern, t)
+            return (
+                str(t).strip(),
+                re.compile(
+                    "\\s+".join(map(re.escape, re.split("\\s+", str(t))))
+                    if ignore_consec_spaces
+                    else re.escape(str(t)),
+                    re.I if ignore_case else 0,
+                ),
             )
-        elif msg is None:
-            msg = ""
+
+        exp_output = [compile_re(t) for t in exp_output]
+        reject_output = [compile_re(t) for t in reject_output]
 
         out = self.run_command(
             command if args is None else ([command] + args),
             input,
             sandboxed=True,
             timeout=timeout,
         )
+
         outcmp = out
+        if highlight_matches and out:
+            for _, r in exp_output:
+                out = r.sub(f"\\033[32m\\g<0>\\033[0m", out)
```

Comment:
Why are these f-strings? It doesn't look like there's anything being substituted here. Would be better to use r-strings, since this means not having to use `\\` as an escape character.

Reply:
Resolved in ea794457e. These are just copied from the `msg` processing where f-strings are needed, and I forgot to replace them. I thought the linter (or black) would take care of this?

Reply:
@jonatanschroeder I thought so too, enabling linting for the grader files required some changes, which included getting rid of unused f-strings in some places. Could you manually run `flake8` on this file and see if any errors get thrown? I'm wondering if the linter is running correctly now.

Reply:
I checked out the commit before your comment, and flake8 does indeed complain about it. `make lint-python` also does, as does `docker/lint_python.sh`. Not sure why this would fail locally, but not in Github CI. @nwalters512 any thoughts?

Reply:
Ok, I believe the issue is that the `graders` directory is not added to the Docker container (see .dockerignore), so the Python linter doesn't find any files in those directories.